File: .skills/openclaw-skills/skills/0xnyk/xint/lib/oauth.ts
/**
 * OAuth 2.0 PKCE authentication for X API v2.
 *
 * Supports:
 * - One-time auth setup with PKCE flow (localhost callback or --manual paste)
 * - Token storage with auto-refresh
 * - Scopes: bookmark/like/follows/list/block/mute read+write, tweet.read/write, users.read, offline.access
 */

import { existsSync, readFileSync, writeFileSync, chmodSync, mkdirSync, renameSync } from "fs";
import { join } from "path";
import { createHash, randomBytes } from "crypto";
import * as readline from "readline";

const SKILL_DIR = join(import.meta.dir, "..");
const TOKENS_PATH = join(SKILL_DIR, "data", "oauth-tokens.json");
const DATA_DIR = join(SKILL_DIR, "data");

const AUTHORIZE_URL = "https://x.com/i/oauth2/authorize";
const TOKEN_URL = "https://api.x.com/2/oauth2/token";
const USERS_ME_URL = "https://api.x.com/2/users/me";
const REDIRECT_URI = "http://127.0.0.1:3333/callback";
const SCOPES =
  "bookmark.read bookmark.write like.read like.write follows.read follows.write list.read list.write block.read block.write mute.read mute.write tweet.read tweet.write users.read offline.access";

// Token expiry buffer — refresh 60s before actual expiration
const EXPIRY_BUFFER_MS = 60_000;

export interface OAuthTokens {
  access_token: string;
  refresh_token: string;
  expires_at: number; // unix ms
  user_id: string;
  username: string;
  scope: string;
  created_at: string;
  refreshed_at: string;
}

// --- PKCE Helpers ---

function base64url(buf: Buffer): string {
  return buf
    .toString("base64")
    .replace(/\+/g, "-")
    .replace(/\//g, "_")
    .replace(/=+$/, "");
}

export function generateCodeVerifier(): string {
  return base64url(randomBytes(32));
}

export function generateCodeChallenge(verifier: string): string {
  const hash = createHash("sha256").update(verifier).digest();
  return base64url(hash);
}

export function generateState(): string {
  return randomBytes(16).toString("hex");
}

// --- Token Storage ---

function ensureDataDir(): void {
  if (!existsSync(DATA_DIR)) {
    mkdirSync(DATA_DIR, { recursive: true });
  }
}

export function loadTokens(): OAuthTokens | null {
  if (!existsSync(TOKENS_PATH)) return null;
  try {
    return JSON.parse(readFileSync(TOKENS_PATH, "utf-8"));
  } catch {
    return null;
  }
}

function saveTokens(tokens: OAuthTokens): void {
  ensureDataDir();
  // Atomic write: write to tmp file then rename
  const tmpPath = TOKENS_PATH + ".tmp";
  writeFileSync(tmpPath, JSON.stringify(tokens, null, 2));
  chmodSync(tmpPath, 0o600);
  renameSync(tmpPath, TOKENS_PATH);
}

// --- Client ID ---

function getClientId(): string {
  if (process.env.X_CLIENT_ID) return process.env.X_CLIENT_ID;

  // Try .env in project directory
  try {
    const envFile = readFileSync(
      join(SKILL_DIR, ".env"),
      "utf-8"
    );
    const match = envFile.match(/X_CLIENT_ID=["']?([^"'\n]+)/);
    if (match) return match[1];
  } catch {}

  throw new Error(
    "X_CLIENT_ID not found. Set it in your environment or in .env"
  );
}

// --- Token Exchange ---

async function exchangeCode(
  code: string,
  codeVerifier: string,
  clientId: string
): Promise<{ access_token: string; refresh_token: string; expires_in: number; scope: string }> {
  const body = new URLSearchParams({
    grant_type: "authorization_code",
    code,
    redirect_uri: REDIRECT_URI,
    client_id: clientId,
    code_verifier: codeVerifier,
  });

  const res = await fetch(TOKEN_URL, {
    method: "POST",
    headers: { "Content-Type": "application/x-www-form-urlencoded" },
    body: body.toString(),
  });

  if (!res.ok) {
    const text = await res.text();
    throw new Error(`Token exchange failed (${res.status}): ${text}`);
  }

  return res.json();
}

async function fetchUserMe(accessToken: string): Promise<{ id: string; username: string }> {
  const res = await fetch(USERS_ME_URL, {
    headers: { Authorization: `Bearer ${accessToken}` },
  });

  if (!res.ok) {
    const text = await res.text();
    throw new Error(`Failed to fetch user info (${res.status}): ${text}`);
  }

  const data = await res.json();
  return { id: data.data.id, username: data.data.username };
}

// --- Refresh ---

export async function refreshTokens(tokens?: OAuthTokens | null): Promise<OAuthTokens> {
  tokens = tokens || loadTokens();
  if (!tokens) {
    throw new Error("No tokens found. Run 'auth setup' first.");
  }

  const clientId = getClientId();

  const body = new URLSearchParams({
    grant_type: "refresh_token",
    refresh_token: tokens.refresh_token,
    client_id: clientId,
  });

  const res = await fetch(TOKEN_URL, {
    method: "POST",
    headers: { "Content-Type": "application/x-www-form-urlencoded" },
    body: body.toString(),
  });

  if (!res.ok) {
    const text = await res.text();
    throw new Error(
      `Token refresh failed (${res.status}): ${text}\n` +
      "You may need to re-run 'auth setup' if the refresh token expired."
    );
  }

  const data = await res.json();
  const now = new Date();

  const updated: OAuthTokens = {
    access_token: data.access_token,
    refresh_token: data.refresh_token, // X rotates refresh tokens
    expires_at: now.getTime() + data.expires_in * 1000,
    user_id: tokens.user_id,
    username: tokens.username,
    scope: data.scope || tokens.scope,
    created_at: tokens.created_at,
    refreshed_at: now.toISOString(),
  };

  saveTokens(updated);
  return updated;
}

// --- Get Valid Token (auto-refresh) ---

export async function getValidToken(): Promise<string> {
  const tokens = loadTokens();
  if (!tokens) {
    throw new Error(
      "No OAuth tokens found. Run 'auth setup' first.\n" +
      "Usage: xint auth setup [--manual]"
    );
  }

  // Check expiration with buffer
  if (Date.now() >= tokens.expires_at - EXPIRY_BUFFER_MS) {
    console.error("OAuth token expired, refreshing...");
    const refreshed = await refreshTokens(tokens);
    console.error(`Token refreshed for @${refreshed.username}`);
    return refreshed.access_token;
  }

  return tokens.access_token;
}

// --- Auth Setup ---

function buildAuthorizeUrl(clientId: string, codeChallenge: string, state: string): string {
  const params = new URLSearchParams({
    response_type: "code",
    client_id: clientId,
    redirect_uri: REDIRECT_URI,
    scope: SCOPES,
    state,
    code_challenge: codeChallenge,
    code_challenge_method: "S256",
  });
  return `${AUTHORIZE_URL}?${params.toString()}`;
}

function prompt(question: string): Promise<string> {
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stderr,
  });
  return new Promise((resolve) => {
    rl.question(question, (answer) => {
      rl.close();
      resolve(answer.trim());
    });
  });
}

export async function authSetup(manual: boolean = false): Promise<void> {
  const clientId = getClientId();
  const codeVerifier = generateCodeVerifier();
  const codeChallenge = generateCodeChallenge(codeVerifier);
  const state = generateState();
  const authorizeUrl = buildAuthorizeUrl(clientId, codeChallenge, state);

  console.error("\n=== X API OAuth 2.0 Setup (PKCE) ===\n");
  console.error("1. Open this URL in your browser:\n");
  console.error(authorizeUrl);
  console.error("\n2. Authorize the app, then:");

  let code: string;

  if (manual) {
    // Manual mode: user pastes the redirect URL
    console.error("   Paste the full redirect URL here:\n");
    const redirectUrl = await prompt("Redirect URL> ");

    const url = new URL(redirectUrl);
    const returnedState = url.searchParams.get("state");
    code = url.searchParams.get("code") || "";

    if (returnedState !== state) {
      throw new Error("State mismatch — possible CSRF attack. Aborting.");
    }
    if (!code) {
      const error = url.searchParams.get("error");
      const desc = url.searchParams.get("error_description");
      throw new Error(`Authorization failed: ${error} — ${desc}`);
    }
  } else {
    // Localhost callback mode
    console.error("   The browser will redirect to localhost:3333\n");

    code = await waitForCallback(state);
  }

  // Exchange code for tokens
  console.error("\nExchanging authorization code for tokens...");
  const tokenData = await exchangeCode(code, codeVerifier, clientId);

  // Fetch user info
  console.error("Fetching user info...");
  const user = await fetchUserMe(tokenData.access_token);

  const now = new Date();
  const tokens: OAuthTokens = {
    access_token: tokenData.access_token,
    refresh_token: tokenData.refresh_token,
    expires_at: now.getTime() + tokenData.expires_in * 1000,
    user_id: user.id,
    username: user.username,
    scope: tokenData.scope,
    created_at: now.toISOString(),
    refreshed_at: now.toISOString(),
  };

  saveTokens(tokens);

  console.error(`\nAuthenticated as @${user.username} (ID: ${user.id})`);
  console.error(`Token expires in ${Math.round(tokenData.expires_in / 60)} minutes`);
  console.error(`Refresh token valid for ~6 months`);
  console.error(`Tokens saved to ${TOKENS_PATH}`);
}

// --- Localhost Callback Server ---

async function waitForCallback(expectedState: string): Promise<string> {
  const TIMEOUT_MS = 5 * 60 * 1000; // 5 minutes

  return new Promise<string>((resolve, reject) => {
    const timeout = setTimeout(() => {
      server.stop();
      reject(new Error("Callback timeout (5 minutes). Try again with --manual flag."));
    }, TIMEOUT_MS);

    const server = Bun.serve({
      port: 3333,
      hostname: "127.0.0.1",
      fetch(req) {
        const url = new URL(req.url);

        if (url.pathname !== "/callback") {
          return new Response("Not found", { status: 404 });
        }

        const returnedState = url.searchParams.get("state");
        const code = url.searchParams.get("code");
        const error = url.searchParams.get("error");

        clearTimeout(timeout);

        // Schedule server shutdown after response
        setTimeout(() => server.stop(), 100);

        if (error) {
          const desc = url.searchParams.get("error_description") || error;
          reject(new Error(`Authorization denied: ${desc}`));
          return new Response(
            `<html><body><h2>Authorization Failed</h2><p>${desc}</p><p>You can close this tab.</p></body></html>`,
            { headers: { "Content-Type": "text/html" } }
          );
        }

        if (returnedState !== expectedState) {
          reject(new Error("State mismatch — possible CSRF attack."));
          return new Response(
            "<html><body><h2>Error: State Mismatch</h2><p>Possible CSRF attack. Please try again.</p></body></html>",
            { headers: { "Content-Type": "text/html" } }
          );
        }

        if (!code) {
          reject(new Error("No authorization code received."));
          return new Response(
            "<html><body><h2>Error</h2><p>No authorization code received.</p></body></html>",
            { headers: { "Content-Type": "text/html" } }
          );
        }

        resolve(code);
        return new Response(
          "<html><body><h2>Authorization Successful!</h2><p>You can close this tab and return to the terminal.</p></body></html>",
          { headers: { "Content-Type": "text/html" } }
        );
      },
    });

    console.error(`Waiting for callback on http://127.0.0.1:3333/callback ...`);
    console.error(`(timeout in 5 minutes — use Ctrl+C or --manual to abort)\n`);
  });
}

// --- Auth Status ---

export function authStatus(): void {
  const tokens = loadTokens();

  if (!tokens) {
    console.log("❌ No OAuth tokens found.");
    console.log("Run: xint auth setup [--manual]");
    return;
  }

  const expiresIn = tokens.expires_at - Date.now();
  const expiresStr = expiresIn > 0
    ? `${Math.round(expiresIn / 60_000)} minutes`
    : "EXPIRED";

  console.log(`✅ Authenticated as @${tokens.username} (ID: ${tokens.user_id})`);
  console.log(`   Scopes: ${tokens.scope}`);
  console.log(`   Access token: ${expiresStr}`);
  console.log(`   Created: ${tokens.created_at}`);
  console.log(`   Last refresh: ${tokens.refreshed_at}`);
}

// --- Auth Refresh (manual) ---

export async function authRefresh(): Promise<void> {
  const tokens = loadTokens();
  if (!tokens) {
    console.error("No tokens found. Run 'auth setup' first.");
    process.exit(1);
  }

  console.error("Refreshing tokens...");
  const refreshed = await refreshTokens(tokens);
  const expiresIn = Math.round((refreshed.expires_at - Date.now()) / 60_000);
  console.log(`✅ Token refreshed for @${refreshed.username}`);
  console.log(`   New token expires in ${expiresIn} minutes`);
}

File: .skills/openclaw-skills/skills/0xnyk/xint-rs/src/auth/oauth.rs
use anyhow::{bail, Result};
use rand::RngCore;
use sha2::{Digest, Sha256};
use std::fs;
use std::io::{self, BufRead, Write as IoWrite};
use std::path::Path;

use crate::client::XClient;
use crate::models::OAuthTokens;

const AUTHORIZE_URL: &str = "https://x.com/i/oauth2/authorize";
const TOKEN_URL: &str = "https://api.x.com/2/oauth2/token";
const REDIRECT_URI: &str = "http://127.0.0.1:3333/callback";
const SCOPES: &str = "bookmark.read bookmark.write like.read like.write follows.read follows.write list.read list.write block.read block.write mute.read mute.write tweet.read tweet.write users.read offline.access";
const EXPIRY_BUFFER_MS: i64 = 60_000;

// ---------------------------------------------------------------------------
// PKCE helpers
// ---------------------------------------------------------------------------

fn base64url(bytes: &[u8]) -> String {
    use base64::Engine;
    base64::engine::general_purpose::URL_SAFE_NO_PAD.encode(bytes)
}

pub fn generate_code_verifier() -> String {
    let mut bytes = [0u8; 32];
    if getrandom::fill(&mut bytes).is_err() {
        rand::thread_rng().fill_bytes(&mut bytes);
    }
    base64url(&bytes)
}

pub fn generate_code_challenge(verifier: &str) -> String {
    let mut hasher = Sha256::new();
    hasher.update(verifier.as_bytes());
    let hash = hasher.finalize();
    base64url(&hash)
}

pub fn generate_state() -> String {
    let mut bytes = [0u8; 16];
    if getrandom::fill(&mut bytes).is_err() {
        rand::thread_rng().fill_bytes(&mut bytes);
    }
    bytes.iter().map(|b| format!("{b:02x}")).collect()
}

// ---------------------------------------------------------------------------
// Token storage
// ---------------------------------------------------------------------------

pub fn load_tokens(tokens_path: &Path) -> Option<OAuthTokens> {
    if !tokens_path.exists() {
        return None;
    }
    let content = fs::read_to_string(tokens_path).ok()?;
    serde_json::from_str(&content).ok()
}

fn save_tokens(tokens_path: &Path, tokens: &OAuthTokens) -> Result<()> {
    if let Some(parent) = tokens_path.parent() {
        fs::create_dir_all(parent)?;
    }
    let tmp = tokens_path.with_extension("json.tmp");
    let json = serde_json::to_string_pretty(tokens)?;
    fs::write(&tmp, &json)?;

    // Set file permissions to 600 on unix
    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt;
        fs::set_permissions(&tmp, fs::Permissions::from_mode(0o600))?;
    }

    fs::rename(&tmp, tokens_path)?;
    Ok(())
}

// ---------------------------------------------------------------------------
// Token exchange / refresh
// ---------------------------------------------------------------------------

async fn exchange_code(
    client: &XClient,
    code: &str,
    code_verifier: &str,
    client_id: &str,
) -> Result<serde_json::Value> {
    client
        .post_form(
            TOKEN_URL,
            &[
                ("grant_type", "authorization_code"),
                ("code", code),
                ("redirect_uri", REDIRECT_URI),
                ("client_id", client_id),
                ("code_verifier", code_verifier),
            ],
        )
        .await
}

async fn fetch_user_me(client: &XClient, access_token: &str) -> Result<(String, String)> {
    let raw = client.oauth_get("users/me", access_token).await?;
    let data = &raw.data;
    let id = data
        .as_ref()
        .and_then(|d| d.get("id"))
        .and_then(|v| v.as_str())
        .ok_or_else(|| anyhow::anyhow!("Failed to get user ID"))?
        .to_string();
    let username = data
        .as_ref()
        .and_then(|d| d.get("username"))
        .and_then(|v| v.as_str())
        .ok_or_else(|| anyhow::anyhow!("Failed to get username"))?
        .to_string();
    Ok((id, username))
}

/// Refresh an existing OAuth token.
pub async fn refresh_tokens(
    client: &XClient,
    tokens_path: &Path,
    client_id: &str,
    tokens: &OAuthTokens,
) -> Result<OAuthTokens> {
    let data = client
        .post_form(
            TOKEN_URL,
            &[
                ("grant_type", "refresh_token"),
                ("refresh_token", &tokens.refresh_token),
                ("client_id", client_id),
            ],
        )
        .await?;

    let now = chrono::Utc::now();
    let expires_in = data
        .get("expires_in")
        .and_then(|v| v.as_i64())
        .unwrap_or(7200);

    let updated = OAuthTokens {
        access_token: data
            .get("access_token")
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .to_string(),
        refresh_token: data
            .get("refresh_token")
            .and_then(|v| v.as_str())
            .unwrap_or(&tokens.refresh_token)
            .to_string(),
        expires_at: now.timestamp_millis() + expires_in * 1000,
        user_id: tokens.user_id.clone(),
        username: tokens.username.clone(),
        scope: data
            .get("scope")
            .and_then(|v| v.as_str())
            .unwrap_or(&tokens.scope)
            .to_string(),
        created_at: tokens.created_at.clone(),
        refreshed_at: now.to_rfc3339(),
    };

    save_tokens(tokens_path, &updated)?;
    Ok(updated)
}

/// Get a valid access token, auto-refreshing if expired.
pub async fn get_valid_token(
    client: &XClient,
    tokens_path: &Path,
    client_id: &str,
) -> Result<(String, OAuthTokens)> {
    let tokens = load_tokens(tokens_path)
        .ok_or_else(|| anyhow::anyhow!("No OAuth tokens found. Run 'auth setup' first."))?;

    let now = chrono::Utc::now().timestamp_millis();

    if now >= tokens.expires_at - EXPIRY_BUFFER_MS {
        eprintln!("OAuth token expired, refreshing...");
        let refreshed = refresh_tokens(client, tokens_path, client_id, &tokens).await?;
        eprintln!("Token refreshed for @{}", refreshed.username);
        let token = refreshed.access_token.clone();
        Ok((token, refreshed))
    } else {
        let token = tokens.access_token.clone();
        Ok((token, tokens))
    }
}

// ---------------------------------------------------------------------------
// Auth setup
// ---------------------------------------------------------------------------

fn build_authorize_url(client_id: &str, code_challenge: &str, state: &str) -> String {
    format!(
        "{}?response_type=code&client_id={}&redirect_uri={}&scope={}&state={}&code_challenge={}&code_challenge_method=S256",
        AUTHORIZE_URL,
        client_id,
        urlencoding_encode(REDIRECT_URI),
        urlencoding_encode(SCOPES),
        state,
        code_challenge
    )
}

fn urlencoding_encode(s: &str) -> String {
    let mut result = String::new();
    for c in s.chars() {
        match c {
            'A'..='Z' | 'a'..='z' | '0'..='9' | '-' | '_' | '.' | '~' => result.push(c),
            ' ' => result.push_str("%20"),
            _ => {
                for byte in c.to_string().as_bytes() {
                    result.push_str(&format!("%{byte:02X}"));
                }
            }
        }
    }
    result
}

/// Interactive auth setup flow.
pub async fn auth_setup(
    client: &XClient,
    tokens_path: &Path,
    client_id: &str,
    manual: bool,
) -> Result<()> {
    let code_verifier = generate_code_verifier();
    let code_challenge = generate_code_challenge(&code_verifier);
    let state = generate_state();
    let authorize_url = build_authorize_url(client_id, &code_challenge, &state);

    eprintln!("\n=== X API OAuth 2.0 Setup (PKCE) ===\n");
    eprintln!("1. Open this URL in your browser:\n");
    eprintln!("{authorize_url}");
    eprintln!("\n2. Authorize the app, then:");

    let code = if manual {
        eprintln!("   Paste the full redirect URL here:\n");
        eprint!("Redirect URL> ");
        io::stderr().flush()?;

        let mut input = String::new();
        io::stdin().lock().read_line(&mut input)?;
        let input = input.trim();

        // Parse the URL to extract code and state
        let (returned_state, code) = parse_callback_url(input)?;
        if returned_state != state {
            bail!("State mismatch — possible CSRF attack. Aborting.");
        }
        code
    } else {
        eprintln!("   The browser will redirect to localhost:3333\n");
        wait_for_callback(&state).await?
    };

    eprintln!("\nExchanging authorization code for tokens...");
    let token_data = exchange_code(client, &code, &code_verifier, client_id).await?;

    let access_token = token_data
        .get("access_token")
        .and_then(|v| v.as_str())
        .ok_or_else(|| anyhow::anyhow!("No access_token in response"))?;

    eprintln!("Fetching user info...");
    let (user_id, username) = fetch_user_me(client, access_token).await?;

    let now = chrono::Utc::now();
    let expires_in = token_data
        .get("expires_in")
        .and_then(|v| v.as_i64())
        .unwrap_or(7200);

    let tokens = OAuthTokens {
        access_token: access_token.to_string(),
        refresh_token: token_data
            .get("refresh_token")
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .to_string(),
        expires_at: now.timestamp_millis() + expires_in * 1000,
        user_id: user_id.clone(),
        username: username.clone(),
        scope: token_data
            .get("scope")
            .and_then(|v| v.as_str())
            .unwrap_or(SCOPES)
            .to_string(),
        created_at: now.to_rfc3339(),
        refreshed_at: now.to_rfc3339(),
    };

    save_tokens(tokens_path, &tokens)?;

    eprintln!("\nAuthenticated as @{username} (ID: {user_id})");
    eprintln!("Token expires in {} minutes", expires_in / 60);
    eprintln!("Refresh token valid for ~6 months");
    eprintln!("Tokens saved to {}", tokens_path.display());

    Ok(())
}

fn parse_callback_url(url_str: &str) -> Result<(String, String)> {
    // Simple URL parameter extraction
    let query = if let Some(pos) = url_str.find('?') {
        &url_str[pos + 1..]
    } else {
        bail!("Invalid redirect URL: no query parameters");
    };

    let mut state = String::new();
    let mut code = String::new();
    let mut error = String::new();

    for pair in query.split('&') {
        let mut parts = pair.splitn(2, '=');
        let key = parts.next().unwrap_or("");
        let value = parts.next().unwrap_or("");
        match key {
            "state" => state = value.to_string(),
            "code" => code = value.to_string(),
            "error" => error = value.to_string(),
            _ => {}
        }
    }

    if !error.is_empty() {
        bail!("Authorization failed: {error}");
    }
    if code.is_empty() {
        bail!("No authorization code in redirect URL");
    }

    Ok((state, code))
}

/// Wait for the OAuth callback on localhost:3333.
async fn wait_for_callback(expected_state: &str) -> Result<String> {
    use tokio::io::{AsyncReadExt, AsyncWriteExt};
    use tokio::net::TcpListener;

    let listener = TcpListener::bind("127.0.0.1:3333").await?;
    eprintln!("Waiting for callback on http://127.0.0.1:3333/callback ...");
    eprintln!("(timeout in 5 minutes — use Ctrl+C or --manual to abort)\n");

    let timeout = tokio::time::timeout(std::time::Duration::from_secs(300), async {
        loop {
            let (mut socket, _) = listener.accept().await?;
            let mut buf = vec![0u8; 4096];
            let n = socket.read(&mut buf).await?;
            let request = String::from_utf8_lossy(&buf[..n]);

            // Extract path from GET request
            let first_line = request.lines().next().unwrap_or("");
            let path = first_line.split_whitespace().nth(1).unwrap_or("/");

            if !path.starts_with("/callback") {
                let response =
                    "HTTP/1.1 404 Not Found\r\nContent-Type: text/html\r\n\r\n<h2>Not Found</h2>";
                let _ = socket.write_all(response.as_bytes()).await;
                continue;
            }

            let query = path.split('?').nth(1).unwrap_or("");
            let mut state = String::new();
            let mut code = String::new();
            let mut error = String::new();

            for pair in query.split('&') {
                let mut parts = pair.splitn(2, '=');
                let key = parts.next().unwrap_or("");
                let value = parts.next().unwrap_or("");
                match key {
                    "state" => state = value.to_string(),
                    "code" => code = value.to_string(),
                    "error" => error = value.to_string(),
                    _ => {}
                }
            }

            if !error.is_empty() {
                let html = format!("<html><body><h2>Authorization Failed</h2><p>{error}</p><p>You can close this tab.</p></body></html>");
                let response = format!(
                    "HTTP/1.1 200 OK\r\nContent-Type: text/html\r\nContent-Length: {}\r\n\r\n{}",
                    html.len(),
                    html
                );
                let _ = socket.write_all(response.as_bytes()).await;
                return Err(anyhow::anyhow!("Authorization denied: {error}"));
            }

            if state != expected_state {
                let html = "<html><body><h2>Error: State Mismatch</h2><p>Possible CSRF attack. Please try again.</p></body></html>";
                let response = format!(
                    "HTTP/1.1 200 OK\r\nContent-Type: text/html\r\nContent-Length: {}\r\n\r\n{}",
                    html.len(),
                    html
                );
                let _ = socket.write_all(response.as_bytes()).await;
                return Err(anyhow::anyhow!("State mismatch — possible CSRF attack."));
            }

            if code.is_empty() {
                let html = "<html><body><h2>Error</h2><p>No authorization code received.</p></body></html>";
                let response = format!(
                    "HTTP/1.1 200 OK\r\nContent-Type: text/html\r\nContent-Length: {}\r\n\r\n{}",
                    html.len(),
                    html
                );
                let _ = socket.write_all(response.as_bytes()).await;
                return Err(anyhow::anyhow!("No authorization code received."));
            }

            let html = "<html><body><h2>Authorization Successful!</h2><p>You can close this tab and return to the terminal.</p></body></html>";
            let response = format!(
                "HTTP/1.1 200 OK\r\nContent-Type: text/html\r\nContent-Length: {}\r\n\r\n{}",
                html.len(),
                html
            );
            let _ = socket.write_all(response.as_bytes()).await;

            return Ok::<String, anyhow::Error>(code);
        }
    });

    match timeout.await {
        Ok(result) => result,
        Err(_) => bail!("Callback timeout (5 minutes). Try again with --manual flag."),
    }
}

/// Print auth status.
pub fn auth_status(tokens_path: &Path) {
    let tokens = match load_tokens(tokens_path) {
        Some(t) => t,
        None => {
            println!("No OAuth tokens found.");
            println!("Run: xint auth setup [--manual]");
            return;
        }
    };

    let now = chrono::Utc::now().timestamp_millis();
    let expires_in = tokens.expires_at - now;
    let expires_str = if expires_in > 0 {
        format!("{} minutes", expires_in / 60_000)
    } else {
        "EXPIRED".to_string()
    };

    println!(
        "Authenticated as @{} (ID: {})",
        tokens.username, tokens.user_id
    );
    println!("   Scopes: {}", tokens.scope);
    println!("   Access token: {expires_str}");
    println!("   Created: {}", tokens.created_at);
    println!("   Last refresh: {}", tokens.refreshed_at);
}
